### PR TITLE
fix(learning): isolate observer project scope

### LINKED ISSUE #2746: continuous-learning-v2: observations routed to global scope when hook payload lacks cwd (silent fallback + observer pinned at launch)
## Summary

`continuous-learning-v2` routes observations to the wrong scope when the hook payload has no usable `cwd`. Two independent defects; both present on `main` as of `edebcc8`+ (verified against HEAD, pushed 2026-08-10).

Symptom: instincts that should be project-scoped land in the global bucket, and a long-lived observer daemon can be pinned to `global` for its entire life regardless of which project the user is working in. In my install, 4 of 26 observations in the current cycle were scoped `global` while 22 correctly carried the project id — non-deterministic, same machine, same session.

## Defect 1 — silent fallback in `hooks/observe.sh` (no `else` branch)

`hooks/observe.sh:117-126`:

```bash
if [ -n "$STDIN_CWD" ] && [ -d "$STDIN_CWD" ]; then
  _GIT_ROOT=$(git -C "$STDIN_CWD" rev-parse --show-toplevel 2>/dev/null || true)
  if [ -n "$_GIT_ROOT" ]; then
    export CLAUDE_PROJECT_DIR="$_GIT_ROOT"
    unset CLV2_NO_PROJECT
  else
    unset CLAUDE_PROJECT_DIR
    export CLV2_NO_PROJECT=1
  fi
fi
# <-- no else: when STDIN_CWD is empty or names a non-existent dir,
#     neither CLAUDE_PROJECT_DIR nor CLV2_NO_PROJECT is set.
```

When the guard fails, project detection falls through to `scripts/detect-project.sh` **priority 2** — a bare `git rev-parse --show-toplevel` against whatever CWD the hook process happened to inherit. Scope then depends on the hook's inherited CWD rather than the session's project, which is non-deterministic.

`detect-project.sh` itself is correct. Sourced from a repo it returns the right values:

```
$ . scripts/detect-project.sh; echo $PROJECT_ID $PROJECT_NAME
<hash12> <project-name>
$ cd ~ && . scripts/detect-project.sh; echo $PROJECT_ID
global
```

That second result is the failure mode reached whenever the `observe.sh` guard is skipped.

### Suggested fix

Make the fallback explicit instead of letting priority 2 guess:

```bash
else
  # No usable cwd in the hook payload: fall back explicitly rather than
  # letting detect-project.sh guess from the inherited CWD.
  if [ -z "${CLAUDE_PROJECT_DIR:-}" ] || [ ! -d "${CLAUDE_PROJECT_DIR:-}" ]; then
    _GIT_ROOT=$(git -C "$PWD" rev-parse --show-toplevel 2>/dev/null || true)
    [ -n "$_GIT_ROOT" ] && export CLAUDE_PROJECT_DIR="$_GIT_ROOT"
  fi
fi
```

## Defect 2 — observer pins scope at launch

`agents/start-observer.sh:203-215` snapshots scope into the daemon env via `nohup env PROJECT_ID=... PROJECT_DIR=... OBSERVATIONS_FILE=... `. An observer spawned while scope resolved to `global` stays global for its whole lifetime and keeps analyzing the global bucket even after the user moves into a project. Observable in `observer.log`:

```
[Sun Aug  9 21:41:47 PDT 2026] Observer started for global (PID: 43475)
[Sun Aug  9 21:51:52 PDT 2026] Analyzing 25 observations for project global...
```

Restarting the observer from inside a repo re-pins it correctly, but that is a workaround. Re-detecting scope per analysis cycle (or keying the analysis off each observation's own `project_id`) would fix it properly.

## Defect 3 (minor, blocks diagnosis) — `cwd` is parsed but never persisted

`observe.sh:234` reads `cwd` into the intermediate `PARSED` blob, but the writer at `observe.sh:307-314` builds the stored record from a fixed key list that omits it:

```python
observation = {
    "timestamp": ..., "event": ..., "tool": ...,
    "session": ..., "project_id": ..., "project_name": ...
}
```

Result: `cwd` appears in **0 of 11,358** stored records across all archives on my install, so there is no way to tell after the fact whether a payload carried `cwd` or not. Persisting it (or logging it once at debug level) would make defect 1 diagnosable from disk instead of requiring a live capture. Note this also means the absence of `cwd` in stored data is *not* evidence that Claude Code omits the field — that question is still open and needs a live payload capture to settle.

## Environment

- ecc `2.0.0-rc.1`, installed commit `edebcc89efa09dc2705748151d98286ff3bb6023`
- macOS (darwin 25.6.0), zsh, bash 3.2 for hooks
- Plugin-registered hooks (`hooks/hooks.json`); user `settings.json` `PreToolUse` is empty
- Repo remote form: an SSH URL with a non-standard port (`ssh://host:2222/path/<name>`); `_clv2_normalize_remote_url` handles it correctly and hashes stably
- 53 instincts (36 project / 17 global); no parse errors in any observation file

## Impact

Low-grade but cumulative: a fraction of observations get analyzed as universal patterns when they are project-specific, which is exactly the cross-project contamination v2.1's scoping was introduced to prevent. Existing instincts on my install are correctly scoped; this is a slow leak, not active corruption.

### PR DESCRIPTION
**Pull request by dajiaohuang**
## What Changed

- make missing or invalid payload `cwd` deterministic: use a valid explicit `CLAUDE_PROJECT_DIR`, otherwise force global scope instead of inheriting the hook process cwd
- persist a scrubbed `cwd` field in every parsed observation for diagnosis
- isolate observer liveness, lazy-start, and SIGUSR1 signaling to the current `PROJECT_DIR`
- add behavioral coverage for subdirectories, missing cwd, explicit project fallback, and stored cwd
- add a cross-platform source contract for project-scoped observer PIDs

## Why This Change

When hook payloads lack a usable cwd, project detection can currently inherit an unrelated process working directory. A live global observer PID also suppresses project-specific observer startup and can be signaled by unrelated project observations. Together these paths allow slow cross-project scope contamination.

The observer process remains intentionally keyed to one project; the fix prevents a global observer from standing in for a project observer rather than mutating daemon scope in place.

Closes #2746

## Testing Done

- [x] Manual testing completed
- [x] Automated focused tests pass locally
- [x] Edge cases considered and tested

Validation performed:

- Git Bash `bash -n` on `observe.sh` and `start-observer.sh`
- `node tests/hooks/observe-subdirectory-detection.test.js` with Git Bash override (8/8)
- `node tests/hooks/continuous-learning-scope-contract.test.js`
- `node tests/hooks/observer-memory.test.js` (31/31)
- `node tests/hooks/observe-nosurvive-warning.test.js` (3/3)
- `node tests/hooks/observe-signal-counter-race.test.js` (2/2)
- `node scripts/ci/validate-skills.js`
- `node scripts/ci/validate-hooks.js`

`shellcheck` is not installed in the local Windows environment; syntax and behavioral checks used Git Bash. CI provides the Linux validation boundary.

## Type of Change

- [x] `fix:` Bug fix
- [ ] `feat:` New feature
- [ ] `refactor:` Code refactoring
- [ ] `docs:` Documentation
- [x] `test:` Tests
- [ ] `chore:` Maintenance/tooling
- [ ] `ci:` CI/CD changes

## Security & Quality Checklist

- [x] No secrets or API keys committed (ghp_, sk-, AKIA, xoxb, xoxp patterns checked)
- [x] JSON files validate cleanly
- [ ] Shell scripts pass shellcheck (not installed locally; `bash -n` passed)
- [x] Pre-commit hooks pass locally (if configured)
- [x] No sensitive data exposed in logs or output
- [x] Follows conventional commits format

## Documentation

- [ ] Updated relevant documentation
- [x] Added comments for complex logic
- [ ] README updated (if needed)

### CODE REVIEW

File: tests/hooks/observe-subdirectory-detection.test.js [hunk 1]
```diff
@@ -87,15 +90,18 @@ function gitInit(dir) {
...
     tool_input: { file_path: 'README.md' },
     tool_response: 'ok',
     session_id: 'session-subdir-test',
-    cwd,
-  });
+  };
+  if (cwd !== undefined) {
+    hookPayload.cwd = cwd;
+  }
+  const payload = JSON.stringify(hookPayload);
```
[the remark this answers is not in the document]
Comment (dajiaohuang) [reply tests/hooks/observe-subdirectory-detection.test.js:100]: Fixed in 395a2300. The test payload now includes cwd through a conditional spread, and the stored observation is constructed once with its scrubbed optional fields.

File: tests/hooks/observe-subdirectory-detection.test.js [hunk 2]
```diff
@@ -211,6 +217,65 @@ test('observe.sh writes project metadata for the git root when cwd is a subdirec
...
+    });
+    assert.strictEqual(result.status, 0, result.stderr);
+
+    const { metadata, projectDir } = readSingleProjectMetadata(homeDir);
+    assert.strictEqual(normalizeComparablePath(metadata.root), normalizeComparablePath(repoDir));
+    const observation = JSON.parse(
+      fs.readFileSync(path.join(projectDir, 'observations.jsonl'), 'utf8').trim()
+    );
+    assert.strictEqual(observation.cwd, '');
+    assert.notStrictEqual(observation.project_id, 'global');
```
[the remark this answers is not in the document]
Comment (dajiaohuang) [reply tests/hooks/observe-subdirectory-detection.test.js:300]: Fixed in 395a2300. The global-scope case now rejects any project directory, and the explicit-project case rejects a parallel global observations file. The focused scope test remains 8/8.

File: tests/hooks/observe-subdirectory-detection.test.js [hunk 3]
```diff
@@ -87,15 +90,18 @@ function gitInit(dir) {
...
     tool_name: 'Read',
     tool_input: { file_path: 'README.md' },
     tool_response: 'ok',
     session_id: 'session-subdir-test',
-    cwd,
-  });
+  };
+  if (cwd !== undefined) {
+    hookPayload.cwd = cwd;
+  }
```
[the remark this answers is not in the document]
Comment (dajiaohuang) [reply tests/hooks/observe-subdirectory-detection.test.js]: Fixed in 395a2300 by constructing the optional cwd property in the initial payload object.

File: skills/continuous-learning-v2/hooks/observe.sh [hunk 4]
```diff
@@ -353,10 +353,17 @@ def scrub(val):
...
+    "timestamp": os.environ["TIMESTAMP"],
+    "event": parsed["event"],
+    "tool": parsed["tool"],
+    "session": parsed["session"],
+    "project_id": os.environ.get("PROJECT_ID_ENV", "global"),
+    "project_name": os.environ.get("PROJECT_NAME_ENV", "global"),
+    "cwd": scrub(parsed.get("cwd", "")),
+    **({"input": scrub(parsed["input"])} if parsed["input"] else {}),
+    **({"output": scrub(parsed["output"])} if parsed["output"] is not None else {})
+}
```
[the remark this answers is not in the document]
Comment (dajiaohuang) [reply skills/continuous-learning-v2/hooks/observe.sh:367]: Fixed in 59be0060. Both project detection and persisted diagnostics now normalize non-string cwd values to an empty string; the new cwd:null regression case passes, with the focused suite now 9/9.

**Comment by chenhz01** [comment]
The deterministic fallback chain (payload cwd → valid `CLAUDE_PROJECT_DIR` → global) closes the scope-contamination path cleanly, and persisting a scrubbed `cwd` on every observation is a good diagnostic. One residual keying issue: `start-observer.sh` builds `PID_FILE="${PROJECT_DIR}/.observer.pid"` from the raw `PROJECT_DIR` string. The same project reached through a symlink, a trailing slash, or a `..` segment yields two different keys — two observers writing the same `observations.jsonl`, each signaling its own PID, and liveness checks that look dead while the twin is alive. Canonicalizing once before keying is cheap:

```sh
PROJECT_DIR="$(cd "$PROJECT_DIR" 2>/dev/null && pwd -P)" || true
```

The scope-contract test could then also pin `pwd -P` canonicalization so a refactor can't drop it. Happy to open a PR with that plus the assertion.
